### PR TITLE
ACDM Evaluation Metrics

### PR DESCRIPTION
**Pull request by rg625**
Changes added: 
- evaluation metrics as per https://arxiv.org/pdf/2309.01745 .
- data loaders now take any number for input frames
- split encoder into history encoder (past N frames) and present encoder (present frame) and added them in the encoding function
- Handling multiple time-series at once now
- History encoder resembles Transformer now, able to handle irregular number of frames
- Training for Re number separately to see if we can recover it
- Added Re number learning with normalization for stability
- Added DDP learning
- Added environment fixed

 Suggestions?

### CODE REVIEW

File: koopman_autoencoder/models/metrics.py
Comment (Etienne-Meunier) [inline koopman_autoencoder/models/metrics.py:1]: Ok that's the part I am not sure : it seems like by doing (x*255).clamp(0, 255) you are discarding values outside 0-1 in your output and ref, but all values should not be in this range right ?

File: koopman_autoencoder/models/trainer.py
Comment (Etienne-Meunier) [inline koopman_autoencoder/models/trainer.py:1]: I am not sure about the way you disable the gradient here, why don't you just use latent.detach() at the input of the re model ?

**Review by Etienne-Meunier — COMMENTED** [review]
Overall looks good, just need to fix / clarify the metric and gradient detach stuff

File: koopman_autoencoder/models/metrics.py
Comment (rg625) [reply koopman_autoencoder/models/metrics.py:1]: Yes, this is a bit tricky! They do the same in the paper so it is a bit confusing, but my guess is that the multiplication maps it to pixel values which the function is used to work with

File: koopman_autoencoder/models/trainer.py
Comment (rg625) [reply koopman_autoencoder/models/trainer.py:1]: We are doing z_preds.detach() when passing it through self.re() in the autoencoder forward. In trainer we are doing 

if re_loss is not None:
            loss.backward(retain_graph=True)
            
basically, if we want to use the same optimizer for the entire model and not 2 different ones, we need to retain the graph to tell the optimizer how these tensors interact with each other. Since z_preds is detached and not present in the total loss, we need to keep the graph. Personally I do not see much of a difference. The current implementation seems more explicit, but it was running with both the retain_graph=True and False, so we have to actually run it and see how it works.

Comment (rg625) [reply koopman_autoencoder/models/trainer.py:1]: You were right! I did not check that properly with larger batch sizes. Retaining the graph is bad, uses a lot more memory. Also, we are doing gradient checkpointing and I am concerned retaining the graph in loss will badly interfere with the checkpointing so it is better to avoid it.

Changed it in the latest commit. Let me know what you think.

File: koopman_autoencoder/models/metrics.py
Comment (Etienne-Meunier) [reply koopman_autoencoder/models/metrics.py:1]: Yeah, maybe just check that in your ref the values are well in this 0-1 range, otherwise it may cause problems in your prediction by discarding part of the input

File: koopman_autoencoder/models/loss.py [hunk 1]
```diff
@@ -149,7 +149,9 @@ def kl_divergence(self, latent_pred: Tensor):
 
     @staticmethod
     def re(input: Tensor, predicted: Tensor):
-        diff = (input - predicted).squeeze(-1)  # now shape: [b, t]
+        diff = (input.view(-1, 1, 1) - predicted).squeeze(
+            -1
+        ) / 1000  # now shape: [b, t]
```
Comment (Etienne-Meunier) [inline koopman_autoencoder/models/loss.py]: You could maybe use einops + shape checking for this block maybe

File: koopman_autoencoder/models/loss.py [hunk 2]
```diff
@@ -167,6 +179,10 @@ def forward(self, x_recon, x_preds, latent_pred, x_true, x_future):
         pred_loss_per_variable = self.rollout_loss(x_preds, x_future)
         # KL loss for latent space
         latent_loss = self.kl_divergence(latent_pred)
+        # Reynolds Number estimation
+        re_loss = (
+            self.re_loss(x_future["Re"], reynolds) if reynolds is not None else None
+        )
 
         # Combine total losses (only total_loss will backpropagate)
         total_loss = (
```
Comment (Etienne-Meunier) [inline koopman_autoencoder/models/loss.py:205]: I am not sure why you call .values() here, for DDP ?

File: koopman_autoencoder/models/trainer.py
Comment (Etienne-Meunier) [reply koopman_autoencoder/models/trainer.py:1]: Yeah I agree that technically retain graph will consume more memory. 

Looking at the code it looks good to me now, just to make sure, the logic should be : 

```python
latent = encoder(input)
recon = decoder(latent)
rey = reynoldnet(latent.detach())
predlat = dynamic(latent)
pred = decoder(predlat)
loss_pred = ||gt - pred||_2^2
loss_recon = ||input - recon||_2^2
loss_rey = ||gt_rey - rey||_2^2
total_loss = loss_pred + loss_recon + loss_rey
```

Because of the detach over latent the gradient from loss_rey should not impact the encoder, I think nothing more. If that's what you do then we are good to go :)

**Review by Etienne-Meunier — COMMENTED** [review]
Addressed comment on detach(), looks good

**Review by ypzpy — APPROVED** [review]
